All NSHM 2022 logic tree sensitivity study code to date

AndrewRidden-Harper opened the pull request:
This pull request includes everything done so far in the NSHM 2022 logic tree sensitivity study. As it is a lot of code, I have provided an outline and some pointers to noteworthy parts below. Much of the text in the wiki is taken from the QCAM poster. 

**Code outline**

**Generating hazard curves:**

- `run_toshi_hazard_post_script.py`  Modifies the standard NSHM 2022 logic tree and runs the GNS package `toshi_hazard_post` with the modified logic trees to generate hazard curves. 

- `run_toshi_hazard_post_helper.py` and `logic_tree_tools.py` contain functions and classes used by `run_toshi_hazard_post_script.py`. 

**Creating figures from generated hazard curves:**

- `plotting_script.py` calls functions in `plotting_functions.py` to create the figures. `plotting_helpers.py` contains functions and classes to help make the figures.

**To help direct your attention, some code that could be useful for future projects:**

- `plotting_functions.make_figure_of_srm_and_gmcm_model_dispersions()` creates a figure of subplots, with a column for each provided location. It automatically adjusts the plot to accommodate any number of columns.

- `plotting_functions.make_figures_of_individual_realizations_for_a_single_logic_tree()` creates a figure of subplots showing individual hazard curves. The plots within a row can have quite different values but should also have the same y-axis limits, so the limits are automatically set from the highest and lowest values plotted in the row.  

- `config.yaml` is used to set parameters for hazard curve generation (following the approach in the nzgmdb)

- `logic_tree_tools.py` has functions for modifying logic trees and semi-automatically generating notes that describe the modifications. The approach of generating notes that describe processing may be useful in future projects.

AndrewRidden-Harper committed:
first dev commit

AndrewRidden-Harper committed:
dev commit

AndrewRidden-Harper committed:
dev commit. Need to fix hash lookup

AndrewRidden-Harper committed:
rewriting plotting as functions

AndrewRidden-Harper committed:
dev auto loop over tree

AndrewRidden-Harper committed:
dev commits

AndrewRidden-Harper committed:
basic 4 cases as simple loop

AndrewRidden-Harper committed:
in progress

AndrewRidden-Harper committed:
bug fix

AndrewRidden-Harper committed:
move funcs to modify_logic_tree_in_python.py

AndrewRidden-Harper committed:
reduced number of absolute paths

AndrewRidden-Harper committed:
setup for overnight run

AndrewRidden-Harper committed:
dev commits

AndrewRidden-Harper committed:
under development

AndrewRidden-Harper committed:
load logic trees from json

AndrewRidden-Harper committed:
last commit using subprocess to run

AndrewRidden-Harper committed:
first working tectonic type selection

AndrewRidden-Harper committed:
insert std_ln row after cov row

AndrewRidden-Harper committed:
hazard curves for all GMCM models

AndrewRidden-Harper committed:
dev updates

AndrewRidden-Harper committed:
preliminary realization outputs

AndrewRidden-Harper committed:
Without new files, can't finish writing realizations

Don't know what is causing this issue but will
just have to revert to having seperate files

AndrewRidden-Harper committed:
fixed slow realization writing bug

AndrewRidden-Harper committed:
plotting realizations in progress

AndrewRidden-Harper committed:
output all id components as cols in dataframe

AndrewRidden-Harper committed:
working on filtering the full logic tree

AndrewRidden-Harper committed:
code for realization comparison plots

AndrewRidden-Harper committed:
tidying up functions

AndrewRidden-Harper committed:
Merge remote-tracking branch 'origin/dev' into dev

AndrewRidden-Harper committed:
dev

AndrewRidden-Harper committed:
dev updates

AndrewRidden-Harper committed:
dev updates

AndrewRidden-Harper committed:
small updates

AndrewRidden-Harper committed:
get dispersion of SRM crust models

AndrewRidden-Harper committed:
rough state srm model plots and interp

AndrewRidden-Harper committed:
changed to update plot for slide

AndrewRidden-Harper committed:
slide for poster showcase

AndrewRidden-Harper committed:
tidying up

AndrewRidden-Harper committed:
dev updates

AndrewRidden-Harper committed:
now plots all source models and interp disp

AndrewRidden-Harper committed:
small update

AndrewRidden-Harper committed:
small update

AndrewRidden-Harper committed:
dev updates

AndrewRidden-Harper committed:
makes source model realiztion plots

This code is a bit messily mixed in with
the code that was used for making rough plots
of the ground motion model realizations

AndrewRidden-Harper committed:
updated plotting SRM and GMCM components

AndrewRidden-Harper committed:
save gmm model plots as png

AndrewRidden-Harper committed:
plotting improvements for poster

AndrewRidden-Harper committed:
tidied up code to run toshi_hazard_post

AndrewRidden-Harper committed:
added docstring to run_with_modified_logic_trees

AndrewRidden-Harper committed:
isort and black

AndrewRidden-Harper committed:
tidied up config.yaml

AndrewRidden-Harper committed:
improving plotting code

AndrewRidden-Harper committed:
tidying up plotting

AndrewRidden-Harper committed:
improved plot for poster

AndrewRidden-Harper committed:
adding plotting resource files

AndrewRidden-Harper committed:
code to make explanation plot for poster

AndrewRidden-Harper committed:
Merge remote-tracking branch 'origin/dev' into dev

AndrewRidden-Harper committed:
small changes to plots

AndrewRidden-Harper committed:
populated requirements.txt

AndrewRidden-Harper committed:
improved the plotting code

generalized the function that makes a figure of subplots of the
individual GMCMs and SRMs to take a list of locations of any
length and adjust the figure so that each column is for a location.
Docstrings still in progress.

AndrewRidden-Harper committed:
renamed custom_input_files to input_files

AndrewRidden-Harper committed:
Merge remote-tracking branch 'origin/dev' into dev

AndrewRidden-Harper committed:
changed run_counter to logic_tree_index other names also improved

AndrewRidden-Harper committed:
tidied up several plotting functions. Before deleting old functions.

AndrewRidden-Harper committed:
broken loading code

AndrewRidden-Harper committed:
make_figure_of_gmcms partially tidied up

AndrewRidden-Harper committed:
tidying up plotting functions

AndrewRidden-Harper committed:
working on dispersion range plot

AndrewRidden-Harper committed:
all aggregate stats plots tidied up

AndrewRidden-Harper committed:
working on tidying up functions to plot individual realizations

AndrewRidden-Harper committed:
plot individual models almost complete

AndrewRidden-Harper committed:
tec reg type

AndrewRidden-Harper committed:
broken

AndrewRidden-Harper committed:
bug with source logic tree component id

AndrewRidden-Harper committed:
improving realiztion plotting code

AndrewRidden-Harper committed:
moved CustomLogicTreePair to run_toshi_hazard_post_helper.py

AndrewRidden-Harper committed:
simplified realization plotting

AndrewRidden-Harper committed:
all output in one directory

AndrewRidden-Harper committed:
finished individual realization plot

AndrewRidden-Harper committed:
all plotting functions tidy and working

AndrewRidden-Harper committed:
deleted unneeded comments and prints from logic_tree_tools.py

AndrewRidden-Harper committed:
ran black and isort on logic_tree_tools.py

AndrewRidden-Harper committed:
isort on files

AndrewRidden-Harper committed:
ran black piecewise on plotting_functions.py

AndrewRidden-Harper committed:
finished Black on plotting_functions.py

AndrewRidden-Harper committed:
changed requirements file to use our forks

AndrewRidden-Harper committed:
Black and isort on all files

AndrewRidden-Harper committed:
wiki draft

AndrewRidden-Harper committed:
fixed small plotting issues

AndrewRidden-Harper committed:
first complete wiki draft

AndrewRidden-Harper committed:
small comment improvements

AndrewRidden-Harper committed:
removed #egg from requirements.txt

AndrewRidden-Harper committed:
fix isort test fail

AndrewRidden-Harper committed:
Added brief usage reference

AndrewRidden-Harper committed:
improved variable names and added missing docs

AndrewRidden-Harper committed:
deleted unused code

AndrewRidden-Harper committed:
improved default output directory name

AndrewRidden-Harper committed:
renaming additional variables

AndrewRidden-Harper committed:
bug fix related to variable renaming

AndrewRidden-Harper committed:
uncommented some example code

AndrewRidden-Harper committed:
ran Black and isort on all files

AndrewRidden-Harper committed:
implementing ruff

AndrewRidden-Harper committed:
Fix ruff issues

AndrewRidden-Harper committed:
trying to fix deptry test

AndrewRidden-Harper committed:
trying to fix deptry issue

AndrewRidden-Harper committed:
installs with pip

AndrewRidden-Harper committed:
trying to fix deptry test fail

AndrewRidden-Harper committed:
trying to fix deptry test fail

AndrewRidden-Harper committed:
deleted GNS packages not directly used in codebase for deptry

AndrewRidden-Harper committed:
try to fix deptry test fail

AndrewRidden-Harper committed:
try to fix deptry test fail

AndrewRidden-Harper committed:
try to fix deptry test fail

AndrewRidden-Harper committed:
dev

AndrewRidden-Harper committed:
installs with pip but not deptry test

AndrewRidden-Harper committed:
bug fix

AndrewRidden-Harper committed:
tidying up and readme improvements

AndrewRidden-Harper committed:
wiki improvements

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+
+    nth_highest : int
+        The nth highest weighted branch to reduce the logic tree to. For example, if n = 1, the logic tree is reduced
+        to the first highest weighted branch. If n = 2, it is reduced to the second-highest weighted branch, etc.
```
```suggestion
        to the highest weighted branch. If n = 2, it is reduced to the second-highest weighted branch, etc.
```

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+    Returns
+    -------
+    modified_logic_tree : SourceLogicTree or GMCMLogicTree
+        The logic tree after being reduced only consisting of the nth highest branch in each branch_set.
```
```suggestion
        The logic tree reduced to the nth highest branch in each branch_set.
```

lispandfound replied:
that is much easier for me to parse at least.

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+    ------
+    ValueError
+        If the branches in the modified_logic_tree do not have valid weights.
+        That is its branches in each branch_set do not sum to 1.0.
```
```suggestion
        If the branches in each branch set modified_logic_tree do not sum to 1.0.
```
Just describe the exact condition that produces the error rather than having an extra sentence.

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+
+    # find the maximum number of branches in any branch set
+    max_num_branches = max(
+        [len(branch_set.branches) for branch_set in logic_tree.branch_sets]
```
```suggestion
        len(branch_set.branches) for branch_set in logic_tree.branch_sets
```

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+            selected_branch = copy.deepcopy(reverse_sorted_branches[0])
+        elif nth_highest > len(reverse_sorted_branches):
+            selected_branch = copy.deepcopy(
+                reverse_sorted_branches[len(reverse_sorted_branches) - 1]
```
```suggestion
                reverse_sorted_branches[-1]
```

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+
+
+def reduce_logic_tree_to_nth_highest_weighted_branch(
+    logic_tree: Union[SourceLogicTree, GMCMLogicTree], nth_highest: int
```
Suggestion: Add a type-variable `LogicTree`  to the `nzshm_model.logic_tree` module that encapsulates this union. This would be as simple as defining:

`LogicTree = Union[SourceLogicTree, GMCMLogicTree]`

in `nzshm_model.logic_tree` and then replacing the import for this module with

`from nzshm_model.logic_tree import LogicTree`

You use the union a lot, and it would be easier to read if it was just `LogicTree`. Moreover, this would make it easier to add a logic tree type to the module later since you wouldn't have to update the union in so many places.

AndrewRidden-Harper replied:
nzshm_model.logic_tree is one of the GNS packages and I would prefer to not change that for simplicity. However, it's a good suggestion so as an alternative, I've added LogicTree = Union[SourceLogicTree, GMCMLogicTree] to logic_tree_tools.py so LogicTree can be used

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+        for branch in branch_set.branches:
+            branch_set_running_sum += branch.weight
+
+        branch_set_summed_weights.append(branch_set_running_sum)
```
`branch_set_summed_weights = np.array([[branch.weight for branch in branch_set] for branch_set in logic_tree.branch_sets]).sum(axis=1)`

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+
+        branch_set_summed_weights.append(branch_set_running_sum)
+
+    if not all(np.isclose(np.array(branch_set_summed_weights), 1.0, rtol=1e-15)):
```
With the above, the `np.array` call here is unnecessary and then you should change this to
`if not np.allclose(branch_set_summed_weights, 1):`

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+    return modified_logic_tree_pair
+
+
+def check_weight_validity(logic_tree: Union[SourceLogicTree, GMCMLogicTree]) -> bool:
```
```suggestion
def check_weight_validity(logic_tree: Union[SourceLogicTree, GMCMLogicTree]) -> None:
```

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+            f"The weights of branches in each branch_set do not sum to 1.0.\nThe summed weights for each branch_set are {branch_set_summed_weights}."
+        )
+
+    return True
```
```suggestion
```

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+            values_list.append(values_as_str)
+
+        # Store the list of string values in the dictionary with the branch set short name as the key
+        values_dict[branch_set.short_name] = values_list
```
Maybe clarify what the variables are doing here.

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+
+def logic_tree_pair_with_selected_tectonic_region_types(
+    initial_logic_tree_pair: CustomLogicTreePair,
+    tectonic_region_type_set: list,
+    which_interface: Optional[str] = None,
+) -> list[CustomLogicTreePair]:
+    """
+    Modifies a logic tree pair to only include branch sets that correspond to the selected tectonic region types.
+
+    Parameters
+    ----------
+    initial_logic_tree_pair : CustomLogicTreePair
+        The initial logic tree pair to modify.
+
+    tectonic_region_type_set : list[str]
+        A list of the selected tectonic region types.
+        Valid tectonic region types are:
+            "Active Shallow Crust",
+            "Subduction Interface",
+            "Subduction Intraslab".
+
+    which_interface : str, default = "HIK_and_PUY"
+        Which subduction interfaces to include.
+        Valid options are:
+           "HIK_and_PUY" which includes HIK_and_PUY the Hikurangi–Kermadec (only_HIK) and Puysegur (only_PUY) subduction zones
+           "only_HIK" which includes only the Hikurangi–Kermadec (only_HIK) subduction zone
+           "only_PUY" which includes only the Puysegur (only_PUY) subduction zone.
```
Ditto here wrt the parameter types.

AndrewRidden-Harper replied:
replaced these strings with enum.StrEnum

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+    return modified_logic_tree_pairs
+
+
+def print_info_about_logic_tree(logic_tree: Union[SourceLogicTree, GMCMLogicTree]):
+    """
+    Prints information about a logic tree.
+
+    This function prints details about the type of logic tree (SourceLogicTree or GMCMLogicTree),
+    the number of branch sets it contains, and information about each branch set.
+
+    Parameters
+    ----------
+    logic_tree : Union[SourceLogicTree, GMCMLogicTree]
+        The logic tree for which information is to be printed.
+    """
+
+    print("")  # Add a blank line for readability
+
+    if isinstance(logic_tree, SourceLogicTree):
+        print("Logic tree is a SourceLogicTree")
+
+    if isinstance(logic_tree, GMCMLogicTree):
+        print("Logic tree is a GMCMLogicTree")
+
+    print(f"Logic tree has {len(logic_tree.branch_sets)} branch sets")
+
+    for branch_set_index, branch_set in enumerate(logic_tree.branch_sets):
+
+        print(
+            f"Branch set index {branch_set_index} has name {branch_set.long_name} ({branch_set.short_name}) and contains {len(branch_set.branches)} branches"
+        )
```
Maybe instead of the `print_info_about_logic_tree`, consider adding a `__repr__` method to the logic tree classes and then the user can just type

```
>>> logic_tree
# output of the info function
```

to get this information in a more Pythonic way?

AndrewRidden-Harper replied:
This is a good suggestion but the logic tree classes already have __repr__ methods in the GNS code that I would prefer to leave unchanged for simplicity

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+
+def get_logic_tree_pairs_for_tectonic_selection(
+    initial_logic_tree_pair: CustomLogicTreePair,
+    tectonic_region_type_sets: list[list[str]],
+    which_interfaces: list[str],
```
Again, enums. Would also simplify the types probably.

lispandfound replied:
And the argument names I would again also change.

AndrewRidden-Harper replied:
Changed the argument name tectonic_region_type_sets to tectonic_region_type_groups to avoid potential confusion with the set type

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+        If more than one tectonic_region_type is included in any tectonic_region_type_set.
+    """
+
+    assert (
+        len(tectonic_region_type_sets[0]) == 1
+    ), "Only one tectonic_region_type can be included in each tectonic_region_type_set passed to this function."
```
Assert is nice, but assertion errors are not particularly specific. I would change this to a 

```
if len(tectonic_region_types_set[0]) == 1:
    raise ValueError(...)
```

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+                branch_values_to_find[branch_param_idx] = possible_param_value
+
+                # This is needed as some string conversions return the text of interest surrounded by ' '
+                search_str = remove_single_quotes(str(branch_values_to_find))
```
```suggestion
                search_str = str(branch_values_to_find).replace("'", "")
```

lispandfound replied:
Then remove the function `remove_single_quotes`

lispandfound commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,972 @@
+            for branch in selected_branches_per_param:
+                branch.weight *= needed_scaling_factor
+
+            assert sum(
+                [branch.weight for branch in selected_branches_per_param]
+            ), "Scaled weights of branches do not sum to 1.0"
```
Is this not what the validate logic tree function checks? If so use that instead. Otherwise, change this to an if + valueerror

AndrewRidden-Harper replied:
The validated logic tree function is very similar but it's not exactly the same. The function's input is a whole logic but here we just have a collection of branches so it's probably simpler to leave it like this

lispandfound commented on plotting_functions.py:
```diff
@@ -0,0 +1,1653 @@
+    plot_fontsize: float = 12.0,
+    plot_lineweight: float = 5.0,
+    location: str = "WLG",
+    im: str = "PGA",
```
A [StrEnum](https://docs.python.org/3/library/enum.html#enum.StrEnum) for this would be preferable to a raw string.

AndrewRidden-Harper replied:
Changed string arguments like these to StrEnum

lispandfound commented on plotting_functions.py:
```diff
@@ -0,0 +1,1653 @@
+        The default size was used for the poster and size of
+        (5.12,4.62) was used for the poster showcase slide.
+
+    Returns
+    -------
+    None
```
```suggestion
```

lispandfound commented on plotting_functions.py:
```diff
@@ -0,0 +1,1653 @@
+    plot_dpi : int, optional
+        The resolution of the plot in dots per inch (default is 500).
+
+    Returns
+    -------
+    None
```
```suggestion
```

lispandfound commented on plotting_functions.py:
```diff
@@ -0,0 +1,1653 @@
+    ### latex formatted strings corresponding to the annotation_ims.
+    ### These could be generated automatically from the annotation_ims
+    ### but they are just hard coded for simplicity.
+    manually_matched_latex_strings = [r"10$^{-2}$", r"10$^{-1}$", r"10$^{0}$"]
```
```suggestion
    manually_matched_latex_strings = ["$10^{-2}$", "$10^{-1}$","$10^{0}$"]
```

lispandfound replied:
these strings do not need to be raw strings because they don to contain any escape characters.

lispandfound commented on plotting_functions.py:
```diff
@@ -0,0 +1,1653 @@
+    results_directory: Union[Path, str],
+    plot_output_directory: Union[Path, str],
+    locations: tuple[str, ...] = ("AKL", "WLG", "CHC"),
+    filter_strs: tuple[str, ...] = ("CRU", "HIK_and_PUY", "SLAB"),
```
Flag?

AndrewRidden-Harper replied:
Flags could work for HIK_and_PUY but I would have had to change some of the condition checks so I've changed all these option values to StrEnum for the benefit of Enums but also a relatively simple implementation

lispandfound commented on plotting_functions.py:
```diff
@@ -0,0 +1,1653 @@
+    ### The GMCM logic tree was reduced to the single highest weighted branch so we can use the SRM model components
+    if "1 (nth) h.w.b." in output_notes["ground_motion_logic_tree_note"]:
+        model_name_short = (
+            output_notes["source_logic_tree_note"]
+            .split(">")[1]
+            .split("[")[1]
+            .strip("[ ] ")
+            + "_"
+            + output_notes["source_logic_tree_note"].split(">")[-2].strip()
```
This should probably be a regex. It's very hard to figure out what you are matching here.

lispandfound commented on plotting_functions.py:
```diff
@@ -0,0 +1,1653 @@
+    ### The source logic tree was reduced to the single highest weighted branch so we can use the gmcm models
+    if "1 (nth) h.w.b." in output_notes["source_logic_tree_note"]:
+        model_name_short = (
+            output_notes["ground_motion_logic_tree_note"]
+            .split(">")[-2]
+            .split("*")[0]
+            .strip(" [ ]")
```
Similarly

lispandfound reviewed (changes requested):
I'll go over the docs on the next pass.

AndrewRidden-Harper committed:
Update logic_tree_tools.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
Update logic_tree_tools.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
Update logic_tree_tools.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
Update logic_tree_tools.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
Update logic_tree_tools.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
Update logic_tree_tools.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
Update logic_tree_tools.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
Update plotting_functions.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
Update plotting_functions.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
Replaced Union[SourceLogicTree, GMCMLogicTree] with LogicTree

AndrewRidden-Harper committed:
added "weighted" to several "highest branch"

AndrewRidden-Harper committed:
updated check_weight_validity docstring

AndrewRidden-Harper committed:
fixed updated check_weight_validity

AndrewRidden-Harper committed:
simplified get_source_branch_parameters_and_values with numpy

AndrewRidden-Harper committed:
Update logic_tree_tools.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
Update plotting_functions.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
deleted remove_single_quote function

AndrewRidden-Harper committed:
small changes in logic_tree_tools.py

sungeunbae commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,919 @@
+        Examples:
+         [["Active Shallow Crust"], ["Subduction Interface"], ["Subduction Intraslab"]]
+         [["Active Shallow Crust", "Subduction Interface"], ["Subduction Intraslab"]]
+         ["Subduction Interface"]
```
is ["Subduction Interface"] a type of list[list[str]] without a problem?

lispandfound replied:
It does not have type `list[list[str]]` it should instead have type `Union[list[str], list[list[str]]]`... Except that is really ugly which is why I suggest that Andrew use `enum.Flag` instead because this is exactly the situation it was designed for.

AndrewRidden-Harper replied:
I agree that it would make this typing simpler if an enum.Flag was used because that one object could contain any number of these tectonic region type combinations. However, using enum.Flag would require more modifications than using enum.StrEnum so I used enum.StrEnum everywhere to save time

AndrewRidden-Harper replied:
I did correct the type hint in the docstring though. Good spotting.

joelridden commented on input_files/all_imt_all_agg_input.toml:
```diff
@@ -0,0 +1,38 @@
+
+[logic_trees]
+model_version = "NSHM_v1.0.4"
+#model_version = "NSHM_v1.0.0"
```
Is it worth cleaning up the commented lines?

AndrewRidden-Harper replied:
cleaned up commented lines

joelridden reviewed (commented):
Not a lot for me to comment about as most of this is pretty custom and unsure how much it'd be used again apart from some of the functions you mentioned which were nice, just a potential tidy up comment that could be nice for the files in that directory.

sungeunbae reviewed (approved):
Epic. Excellent  documentation (how to install, how to run) and images for the wiki page.

It will be a personal preference, but adding folders `lib` and `script` and placing .py files to them could be considered?
Then *help.py can be placed in lib folder without the _helper suffix and *script.py can be placed in `script` folder without _script suffix. 

Again, just my personal preference, and I'm happy to approve as it is.

AndrewRidden-Harper committed:
use regex to get useful parts of notes

AndrewRidden-Harper committed:
improved error message

claudio525 commented on config.py:
```diff
@@ -0,0 +1,48 @@
+import yaml
+
+
+class Config:
```
`Config` is quite generic, I personally would use something like `RunConfig` just to be a little bit more specific as to what its for.

claudio525 commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,920 @@
+        allowed values for each parameter.
+    """
+
+    values_dict = {}
```
I'd avoid generic names like this and `values_list`, they make it hard to read the code. Make it as specific as possible, without making a mega name and overthinking it (easier said than done to be fair).

AndrewRidden-Harper replied:
That's a good point. I've renamed them

claudio525 commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,920 @@
+        modified_ground_motion_logic_tree
+    )
+
+    modified_logic_tree_pairs.append(modified_logic_tree_pair)
```
Am I missing something, or will `modified_logic_tree_pairs` only ever contain one value?

AndrewRidden-Harper replied:
I think you're right. Good spotting. I think I must have changed my plans for this function halfway through and forgot to undo the list appending part

claudio525 commented on logic_tree_tools.py:
```diff
@@ -0,0 +1,920 @@
+            branch_values_to_find = copy.deepcopy(hwb_values)
+
+            for possible_param_value in possible_values_for_this_param_index:
+
```
You seem to have a lot single lines broken into their own lines. I'm big fan of breaking code into logical blocks, however, also don't want to overdo it. E.g. there is no need for a new line here imo.

claudio525 replied:
Don't worry about fixing this, its just a comment for the future. This is "research" code, so not worth the effort.

AndrewRidden-Harper replied:
I removed a few blank lines but didn't try to get them all

claudio525 commented on plotting_helpers.py:
```diff
@@ -0,0 +1,728 @@
+    return s.translate(translation_table)
+
+
+def load_aggregate_stats_for_one_logic_tree_one_location(
```
Are these loading functions specifically for plotting?

AndrewRidden-Harper replied:
Moved the loading functions to their own module

claudio525 commented on plotting_helpers.py:
Given the functions in this file, not sure if `plotting_helpers.py` is an appropriate name?

AndrewRidden-Harper replied:
Moved the loading functions to their own module

claudio525 reviewed (commented):
Nice work!

AndrewRidden-Harper committed:
PR suggestions in progress

claudio525 commented on logic_tree_tools.py:
```diff
@@ -289,6 +298,8 @@ def select_branch_sets_given_tectonic_region_type(
     if isinstance(tectonic_region_type_set, str):
         tectonic_region_type_set = [tectonic_region_type_set]
 
+    print()
```
Random print?

AndrewRidden-Harper replied:
sorry, this was probably for a debugging break point

claudio525 commented on logic_tree_tools.py:
```diff
@@ -592,10 +605,12 @@ def get_logic_tree_pairs_for_individual_ground_motion_models(
-        raise ValueError(
-            "Only one tectonic_region_type can be included in each tectonic_region_type_set passed to this function."
-        )
+    print()
+
+    # if len(tectonic_region_type_sets[0]) == 1:
+    #     raise ValueError(
+    #         "Only one tectonic_region_type can be included in each tectonic_region_type_set passed to this function."
+    #     )
```
Can probably tidy this up

AndrewRidden-Harper replied:
sorry, this was all commented out for debugging. Uncommented now.

claudio525 commented on run_toshi_hazard_post_script.py:
```diff
@@ -76,19 +77,19 @@
-        ground_motion_logic_tree_nth_highest=1,
-    ),
-]
+# logic_tree_pair_list1 = [
+#     full_logic_tree_pair,
+#     logic_tree_tools.reduce_logic_tree_pair_to_nth_highest_branches(
+#         full_logic_tree_pair,
+#         source_logic_tree_nth_highest=1,
+#         ground_motion_logic_tree_nth_highest=None,
+#     ),
+#     logic_tree_tools.reduce_logic_tree_pair_to_nth_highest_branches(
+#         full_logic_tree_pair,
+#         source_logic_tree_nth_highest=None,
+#         ground_motion_logic_tree_nth_highest=1,
+#     ),
+# ]
```
Delete?

AndrewRidden-Harper replied:
sorry, this was all commented out for debugging. Uncommented now.

claudio525 commented on run_toshi_hazard_post_script.py:
```diff
@@ -120,15 +121,21 @@
 )
 
 ### concatenate the logic_tree_pair_lists
+# logic_tree_pair_list = (
+#     logic_tree_pair_list1 + logic_tree_pair_list2 + logic_tree_pair_list3
+# )
```
Temporary code?

AndrewRidden-Harper replied:
sorry, this was all commented out for debugging. Uncommented now.

AndrewRidden-Harper committed:
introduced bug fix

AndrewRidden-Harper committed:
fixed one introduced bug

AndrewRidden-Harper committed:
working with StrEnum but old docs

AndrewRidden-Harper committed:
updated docstrings for StrEnum

AndrewRidden-Harper committed:
Merge branch 'try_enum_V2' into dev

# Conflicts:
#	logic_tree_tools.py
#	run_toshi_hazard_post_script.py

AndrewRidden-Harper committed:
replaced argument str with StrEnum

AndrewRidden-Harper committed:
first set of regular expression simplified

AndrewRidden-Harper committed:
simplified regular expressions

AndrewRidden-Harper committed:
PR suggestions

AndrewRidden-Harper committed:
refactored code following PR suggestions

AndrewRidden-Harper committed:
more refactoring for PR suggestions

AndrewRidden-Harper committed:
renaming and refactoring

AndrewRidden-Harper committed:
regular expressions improved

AndrewRidden-Harper committed:
renamed values_list to source_branch_parameter_values

AndrewRidden-Harper committed:
wiki updates

AndrewRidden-Harper committed:
logic_tree_pair_with_selected_tectonic_region_types returns single value

before it was returning a list of logic_tree_pairs that apparently
could only have one value

AndrewRidden-Harper committed:
tried to refactor as a package

AndrewRidden-Harper committed:
tried to refactor as a packaged (adding)

AndrewRidden-Harper committed:
working version after refactoring as a package

AndrewRidden-Harper committed:
updated wiki with new installation and usage instructions

AndrewRidden-Harper committed:
completed refactoring as a package

AndrewRidden-Harper committed:
removed unused "config" from requirements.txt

AndrewRidden-Harper committed:
added Ruff ignore # noqa: N815 for only_HIK

AndrewRidden-Harper committed:
fixed small plotting issue

AndrewRidden-Harper commented:
Thanks for all your comments and suggestions. I've finished incorporating them so it's ready for re-review. Note that I restructured the code and now it all looks like new code in the GitHub diff so I'm sorry about that

joelridden commented on nshm_logic_tree_utilities/lib/param_options.py:
I'm unsure of the naming of this file, I first thought it was more of a config that I could adjust for the parameters. I see it more of a constants as it's also used in code logic not just parameter definitions for strings etc.

AndrewRidden-Harper replied:
That's a good point. I've renamed param_options.py to constants.py

joelridden commented on wiki/Home.md:
```diff
@@ -13,29 +13,29 @@ With this package, you can create modified logic trees and use them to calculate
+
+   1. Go to the `nshm_logic_tree_utilities/scripts` directory (where this `nshm_logic_tree_utilities` directory is inside the first)
+   2. Open `nshm_logic_tree_utilities/run_toshi_hazard_post_config.yaml`
+   3. Modify the line containing `output_directory : "/home/arr65/data/nshm/output"` to use your directory
```
More for future, I think it can be better to instead of a file someone has to change to specify the output for a script to have this as a script argument input for an output directory.

AndrewRidden-Harper replied:
That's a good suggestion. I'll leave this using a config file for now (just to save time) but if I do more on this project, I might change it to use a script argument input. I'll also keep this in mind for future projects

joelridden reviewed (commented):
Both comments up for discussion / debate

lispandfound commented on wiki/Home.md:
Rather than have this in the wiki, I think this is a `README.md` file. You're giving instructions on how to use the package which someone should see immediately when they land on the repo. Maybe change this location to the top-level of the repo.

lispandfound commented on nshm_logic_tree_utilities/lib/loading_functions.py:
```diff
@@ -0,0 +1,293 @@
+
+def load_aggregate_stats_for_one_logic_tree_one_location(
+    results_dir_for_one_logic_tree: Union[Path, str], location: str
+) -> pd.DataFrame:
```
Also put the location codes here...

lispandfound replied:
and everywhere else in this module you use location codes make the type your new enum.

lispandfound commented on nshm_logic_tree_utilities/lib/loading_functions.py:
```diff
@@ -0,0 +1,293 @@
+    results_df = pd.DataFrame()
+
+    if location not in ["AKL", "WLG", "CHC"]:
+        raise ValueError("location must be AKL, WLG or CHC")
```
...then you can remove this line

lispandfound commented on nshm_logic_tree_utilities/lib/loading_functions.py:
```diff
@@ -0,0 +1,293 @@
+    if location == "WLG":
+        nloc_str = "nloc_0=-41.0~175.0"
+    if location == "AKL":
+        nloc_str = "nloc_0=-37.0~175.0"
```
this should be if/elif/elsee

lispandfound commented on nshm_logic_tree_utilities/lib/loading_functions.py:
```diff
@@ -0,0 +1,293 @@
+    if location == "AKL":
+        nloc_str = "nloc_0=-37.0~175.0"
+
+    # noinspection PyUnboundLocalVariable
```
The above makes this unnecessary

lispandfound commented on nshm_logic_tree_utilities/lib/loading_functions.py:
```diff
@@ -0,0 +1,293 @@
+    """
+
+    data_df: pd.DataFrame()
+    collated_notes_df: pd.DataFrame()
```
```suggestion
    data_df: pd.DataFrame
    collated_notes_df: pd.DataFrame
```

lispandfound commented on nshm_logic_tree_utilities/lib/loading_functions.py:
```diff
@@ -0,0 +1,293 @@
+
+def lookup_realization_name_from_hash(
+    individual_realization_df: pd.DataFrame, registry_directory: Union[Path, str]
+) -> list[plotting_utilities.RealizationName]:
```
American spelling of Realization here.

AndrewRidden-Harper replied:
That's a good point but I think I will leave it as this spelling because I think I consistently used this spelling so if I try to change them all, I might miss some, and having inconsistent spelling might be worse

lispandfound commented on nshm_logic_tree_utilities/lib/logic_tree_tools.py:
```diff
@@ -0,0 +1,873 @@
+    ):
+        return modified_logic_tree_pair
+
+    if source_logic_tree_nth_highest is not None:
```
```suggestion
    if source_logic_tree_nth_highest and not ground_motion_logic_tree_nth_highest:
        return modified_logic_tree_pair

    if source_logic_tree_nth_highest:
```

lispandfound commented on nshm_logic_tree_utilities/lib/logic_tree_tools.py:
```diff
@@ -0,0 +1,873 @@
+        If the weights of branches in each branch_set do not sum to 1.0.
+    """
+
+    logic_tree = copy.deepcopy(logic_tree)
```
Do you really need a deepcopy here? You aren't modifying the object at all.

lispandfound commented on nshm_logic_tree_utilities/lib/logic_tree_tools.py:
```diff
@@ -0,0 +1,873 @@
+        pass
+    else:
+        # remove correlations
+        modified_logic_tree.correlations = LogicTreeCorrelations()
```
```suggestion
    if (param_options.InterfaceName.only_PUY not in branch_set_short_names) | (
        param_options.InterfaceName.only_HIK not in branch_set_short_names
    ):
        # retain the HIK to PUY correlations
        # remove correlations
        modified_logic_tree.correlations = LogicTreeCorrelations()
```

AndrewRidden-Harper replied:
branch_set_short_names is just a list like ["CRU", "HIK", "PUY"] that comes from the GNS code so this can be as simple as:
    `if not (("HIK" in branch_set_short_names) and ("PUY" in branch_set_short_names)):`
    `    # HIK and PUY are not both in the logic tree so remove their correlations`
   `    modified_logic_tree.correlations = LogicTreeCorrelations()`

lispandfound commented on nshm_logic_tree_utilities/lib/loading_functions.py:
```diff
@@ -0,0 +1,293 @@
+        ).split(", ")
+
+        for contributing_branches_hash_id in contributing_branches_hash_ids_clean:
+            seismicity_rate_model_id = contributing_branches_hash_id[0:12]
+            ground_motion_characterization_models_id = contributing_branches_hash_id[
+                12:24
+            ]
```
Comment the structure of the hash id so that these indices make sense.

lispandfound commented on nshm_logic_tree_utilities/lib/logic_tree_tools.py:
```diff
@@ -0,0 +1,873 @@
+    """
+    short_tectonic_region_type_lookup_dict = toml.load(
+        Path(__file__).parent.parent
+        / "resources/short_tectonic_region_type_lookup.toml"
```
```suggestion
        / "resources" / "short_tectonic_region_type_lookup.toml"
```

lispandfound commented on nshm_logic_tree_utilities/lib/logic_tree_tools.py:
```diff
@@ -0,0 +1,873 @@
+
+        assert (
+            len(source_branch_set_names_to_change) == 1
+        ), "Should only have one source branch set to change at this point"
```
ValueError

AndrewRidden-Harper commented on nshm_logic_tree_utilities/lib/logic_tree_tools.py:
```diff
@@ -0,0 +1,873 @@
+
+    modified_logic_tree = copy.deepcopy(logic_tree)
+    new_branch_sets = []
+    for branch_set in logic_tree.branch_sets:
```
Try restructuring this loop and logic checks like in Jake's suggestion here:

`new_branch_sets = [
        copy.deepcopy(branch_set)
        for branch_set in logic_tree.branch_sets
        if branch_set.tectonic_region_type in tectonic_region_type_group
        or (isinstance(logic_tree, SourceLogicTree) and
            tectonic_region_type == param_options...
            (which_interface == param_options.InterfaceName.HIK_and_PUY or which_interface[len('only_'):] == branch_set.short_name))
]`

AndrewRidden-Harper committed:
deleted debug print

AndrewRidden-Harper committed:
renamed param_options.py to constants.py

AndrewRidden-Harper committed:
moved installation and usage instructions to README.md

AndrewRidden-Harper committed:
finished refactoring after param_options.py was renamed to constants.py

AndrewRidden-Harper committed:
changed location str to constant.LocationCode

AndrewRidden-Harper committed:
Update nshm_logic_tree_utilities/lib/loading_functions.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
Update nshm_logic_tree_utilities/lib/logic_tree_tools.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
changed an assert to instead raise ValueError

AndrewRidden-Harper committed:
removed unneeded deepcopy

AndrewRidden-Harper committed:
added comment explaining the logic tree hash ids

AndrewRidden-Harper committed:
simplified check for removing HIK and PUY correlations

AndrewRidden-Harper committed:
small print improvement

AndrewRidden-Harper committed:
Merge remote-tracking branch 'origin/dev' into dev

# Conflicts:
#	nshm_logic_tree_utilities/lib/logic_tree_tools.py

AndrewRidden-Harper committed:
Update nshm_logic_tree_utilities/lib/logic_tree_tools.py

Co-authored-by: Jake Faulkner <jakefaulkn@gmail.com>

AndrewRidden-Harper committed:
small correction to suggested change on GitHub pull request

 if [not] source_logic_tree_nth_highest and not ground_motion_logic_tree_nth_highest:
        return modified_logic_tree_pair

AndrewRidden-Harper committed:
attempt of big if block list append as a list comprehension

AndrewRidden-Harper committed:
added a temporary check that new_branch_sets is the same
